### PR TITLE
Version 0.3.3

### PR DESCRIPTION
**Pull request by oddaspa**
# [0.3.3] - 2022-04-25

## Changed
- Renamed file `convert.py` to `time.py`.
- Refactored `pagination.GetDates` and changed name to `pagination.TimeIterator`.
- Refactored `pagination.GetItems` and changed name to `pagination.ItemIterator`.
- Minor changes to names in `time.py` to be more precise.
- Refactored `@iterator` wrapper to use said iterators.
- `pyclarify.client.make_request()` now uses iterator and is only method for sending requests.
- Extended `models.response.Response` to handle add operation.
- Refactored return on all methods in `client.ClarifyClient` and `client.APIClient`.

## Removed
 - `pyclarify.__utils__.convert.str_to_datetime()`
 - `pyclarify.client.pretty_response()`
 - `pyclarify.client.iterate_bool()`
 - `pyclarify.client.send_simple()`
 - `pyclarify.client.send_iter()`

## Added
- RFC3339 to time delta function in time.py.
- `__add__()` functionality to SelectItemsResponse model and GenericResponse model.
- Filter model to create filters and combining them.
  Filter is based on MongoDB filters and works as described in [the docs](https://docs.clarify.io/api/1.1beta1/methods/filter-syntax).

  Usage:
  ```python
  from pyclarify import filter

  f1 = filter.Filter(fields={"name": filter.NotEqual(value="Lufttemperatur")})
  f2 = filter.Filter(fields={"labels.unit-type": filter.NotIn(value=["Flåte", "Merde 5"])})

  f1.to_query()
  >>> {'name': {'$ne': 'Lufttemperatur'}}

  f3 = f1 & f2
  f3.to_query()
  >>> {
  >>>     '$and': [
  >>>         {'name': {'$ne': 'Lufttemperatur'}},
  >>>         {'labels.unit-type': {'$nin': ['Flåte', 'Merde 5']}}
  >>>     ]
  >>>}
  ```
  Complete list of operators:
  - Equal
  - NotEqual
  - Regex
  - In
  - NotIn
  - LessThan
  - GreaterThan
  - GreaterThanOrEqual
- Exception class for `FilterError` and `TypeError` both located in `__utils__.exceptions`.

## Fixed
- Validator on `models.data.DataFrame` now converts `numpy.nan` to native python `none` .
- `annotations` field on `models.data.SignalInfo` is now optional.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -133,4 +133,7 @@ dmypy.json
 
 # Credentials
 clarify-credentials.json
-clarify-credentials-dev.json
\ No newline at end of file
+clarify-credentials-dev.json
+
+# test files
+test.py
```
Comment (smyrman) [inline .gitignore:139]: Missing trailing newline. 

Suggestion: install/enable editorconfig plugin for your editor: https://editorconfig.org/
Include config at repo root for tab indentation and trailing newlines.

See other repos, e.g. the clarify repo.

File: src/pyclarify/__init__.py [hunk 2]
```diff
@@ -17,3 +17,4 @@
 
 from pyclarify.client import APIClient, ClarifyClient
 from pyclarify.models.data import Signal, SignalInfo, Item, DataFrame
+import pyclarify.__utils__.filter as filter
```
Comment (smyrman) [inline src/pyclarify/__init__.py:20]: Would it perhaps be easier to better to have a folder `pyclarify/filter` and define a file:
`pyclarify/filter/__init__.py`?

Example from Django (not super relevant):

- https://github.com/django/django/blob/main/django/db/models/__init__.py
- https://github.com/django/django/blob/main/django/db/__init__.py

File: src/pyclarify/__utils__/exceptions.py [hunk 3]
```diff
@@ -31,11 +31,58 @@ class PyClarifyImportError(PyClarifyException):
...
+            message or "The functionality your are trying to use requires '{}' to be installed.".format(
                 self.module
             )
         )
 
     def __str__(self):
         return self.message
+
+
+class PyClarifyFilterError(PyClarifyException):
```
Comment (smyrman) [inline src/pyclarify/__utils__/exceptions.py:43]: Why not just `FilterError`?

File: src/pyclarify/__utils__/exceptions.py [hunk 4]
```diff
@@ -31,11 +31,58 @@ class PyClarifyImportError(PyClarifyException):
...
+                self.desired_type,
+                self.actual_values
+            )
+        )
+
+    def __str__(self):
+        return self.message
+
+             
+class PyClarifyTypeError(PyClarifyException):
```
Comment (smyrman) [inline src/pyclarify/__utils__/exceptions.py:69]: Why not just `TypeError`?

File: src/pyclarify/__utils__/filter.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+limitations under the License.
+"""
+from .exceptions import PyClarifyFilterError
+from pydantic.class_validators import root_validator
+from pydantic import BaseModel, Extra
+from pydantic.fields import Optional
+from typing import ForwardRef, Union, List, Dict
+from enum import Enum
+
+class LegalOperators(str, Enum):
```
Comment (smyrman) [inline src/pyclarify/__utils__/filter.py:23]: Allowed is better then Legal. but should the type actually just be `Operator`?

A better name still is actually Comperator.

Comment (smyrman) [reply src/pyclarify/__utils__/filter.py:23]: Naming example from python.com (they just call it Color -- singular):
- https://docs.python.org/3/library/enum.html#creating-an-enum

File: src/pyclarify/__utils__/filter.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    NE = "$ne"
+    REGEX = "$regex"
+    IN = "$in"
+    NIN = "$nin"
+    LT = "$lt"
+    GT = "$gt"
+    GTE = "$gte"
+
+
+class Comparison(BaseModel):
```
Comment (smyrman) [inline src/pyclarify/__utils__/filter.py:33]: Technically, a Comparison could implement multiple operators....

Are we able to expand to that later? I.e. the Go equivalent is something like:

```py
class Comparison(BaseModel):
  in: Union[None,List[str],List[int],List[float],List[datetime]]
  not_in: Union[None,List[str],List[int],List[float],List[datetime]]
  greater: Union[None,str,int,float,datetime]
  greater_or_equal: Union[None,str,int,float,datetime]
  less: Union[None,str,int,float,datetime]
  less_or_equal: Union[None,str,int,float,datetime]
```

File: src/pyclarify/models/data.py [hunk 7]
```diff
@@ -157,18 +162,15 @@ class SignalInfo(BaseModel):
     type: TypeSignal = TypeSignal.numeric
     description: str = ""
     labels: Dict[LabelsKey, List[str]] = {}
-    annotations: Dict[AnnotationKey, str] = {}
+    annotations: Optional[Dict[AnnotationKey, str]] = {}
```
Comment (smyrman) [inline src/pyclarify/models/data.py:165]: For SaveSignals, everything is optional... omitted fields will be set to the _default_ value.

For PublishSignals (item model), there is a set of required fields, including `name`, `type` and `sourceType` I think...

File: src/pyclarify/models/data.py [hunk 8]
```diff
@@ -157,18 +162,15 @@ class SignalInfo(BaseModel):
     type: TypeSignal = TypeSignal.numeric
     description: str = ""
     labels: Dict[LabelsKey, List[str]] = {}
```
Comment (smyrman) [inline src/pyclarify/models/data.py:164]: Possible to use this?
```suggestion
    labels: Dict[LabelsKey, Set[str]] = {}
```

Comment (oddaspa) [reply src/pyclarify/models/data.py:164]: Using set changes the labels from:
```python
labels={'depth': [], 'locality': ['Langøy'], 'site': ['Langøy'], 'topic': ['efb/LA/FL/AT/1'], 'unit-type': ['Flåte'], 'unit_name': ['Flåte']}
```
to
```python
labels={'depth': set(), 'locality': {'Langøy'}, 'site': {'Langøy'}, 'topic': {'efb/LA/FL/AT/1'}, 'unit-type': {'Flåte'}, 'unit_name': {'Flåte'}}
```

I belive that the current version is inline with the documentation.

File: src/pyclarify/models/data.py (hunk 7 above)
Comment (oddaspa) [reply src/pyclarify/models/data.py:165]: Will leave as is for now and update when we switch to the view models.

File: src/pyclarify/__utils__/filter.py (hunk 6 above)
Comment (oddaspa) [reply src/pyclarify/__utils__/filter.py:33]: Yes, we should support multiple operators. Will change in the future.

File: .gitignore (hunk 1 above)
Comment (oddaspa) [reply .gitignore]: Installed editorconfig 👍

File: src/pyclarify/__init__.py (hunk 2 above)
Comment (oddaspa) [reply src/pyclarify/__init__.py]: Changed from `__utils__` for `query`.

File: src/pyclarify/__utils__/exceptions.py (hunk 4 above)
Comment (oddaspa) [reply src/pyclarify/__utils__/exceptions.py]: Now called TypeError.

File: src/pyclarify/__utils__/filter.py (hunk 5 above)
Comment (oddaspa) [reply src/pyclarify/__utils__/filter.py]: Kept as Operator, as in _Logical Operator_.

File: src/pyclarify/models/data.py (hunk 8 above)
Comment (smyrman) [reply src/pyclarify/models/data.py:164]: Array works.
Just be aware that we _treat_ it as a set on the server side, which means that if you submit `["foo", "foo", "bar"]`, we store `["bar", "foo"]`

File: .gitignore [hunk 9]
```diff
@@ -133,4 +133,10 @@ dmypy.json
...
+
+# Editor
+.editorconfig
```
Comment (smyrman) [inline .gitignore:142]: You are supposed to add .editorconfig to the repo

File: src/pyclarify/query/__init__.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+"""
+Copyright 2021 Clarify
```
Comment (smyrman) [inline src/pyclarify/query/__init__.py:2]: 2022 (or 2021-2022 for existing files).

Should it be "Searis AS" instead of "Clarify" for legal reasons?

**Review by smyrman — APPROVED** [review]
Some comments, otherwise looks good.

**Pull request merged**